src/librustc_mir/util/liveness.rs

```diff
@@ -191,33 +201,14 @@ impl LivenessResult {
                 block,
                 statement_index,
             };
-            let statement_defs_uses = self.defs_uses(mir, statement_location, statement);
-            statement_defs_uses.apply(&mut bits);
+            visitor.defs_uses.clear();
+            statement.apply(statement_location, &mut visitor);
+            visitor.defs_uses.apply(&mut bits);
```

Comment:
Hmm. It'd be nice to encapsulate the "clear, apply, etc" logic into some helper so as to make this more obvious. Since we only want to clear the bits before each statement, though, perhaps some setup like this would be good:

```rust
impl DefsUseVisitor {
    fn clear_then_update_bits(&mut self, location: Location, value: &impl MirVisitable<'tcx>, bits: &mut LocalSet) {
        self.clear();
        self.update_bits(location, value, bits);
    }

    /// Update `bits` with the affects of `value`. We should always visit in reverse order.
    /// This method assumes that we have not visited anything before; if you have, use `clear_and_update_bits`.
    fn update_bits(&mut self, location: Location, value: &impl MirVisitable<'tcx>, bits: &mut LocalSet) {
        value.apply(location, self);
        self.defs_uses.apply(bits);
    }
}
```

What do you think?

Reply:
TBH, think it's less clear, and it's more code, and those factors outweigh the benefit of not repeating. But I'll do it if that's what is necessary for an r+ :)

Reply:
Yes, it would please me — not necessarily those exact names and setup, but some kind of method that groups together the logical operation that is occuring here ("clear flag, do visit, apply the result"), and ideally some comments about what the heck is going on.

-- Your capricious reviewer